File: Games/Uzta/src/main/kotlin/jokatu/game/games/uzta/graph/UztaGraph.java
package jokatu.game.games.uzta.graph;

import ophelia.graph.BiGraph;
import org.jetbrains.annotations.NotNull;

import java.util.Optional;

import static java.util.Optional.empty;

public interface UztaGraph extends BiGraph<Node, LineSegment> {

	@NotNull
	default Optional<Node> getNode(@NotNull String id) {
		return getNodes().stream()
				.filter(node -> id.equals(node.getId()))
				.findAny();
	}

	@NotNull
	default Optional<LineSegment> getEdge(@NotNull Node start, @NotNull Node end) {
		if (start.equals(end)) {
			return empty();
		}
		return getEdges().stream()
				.filter(edge -> edge.contains(start))
				.filter(edge -> edge.contains(end))
				.findAny();
	}
}

File: Games/Uzta/src/main/kotlin/jokatu/game/games/uzta/graph/UztaGraph.kt
package jokatu.game.games.uzta.graph

import ophelia.kotlin.graph.UndirectedGraph
import java.util.*
import java.util.Optional.empty

interface UztaGraph : UndirectedGraph<Node, LineSegment> {

	fun getNode(id: String): Optional<Node> {
		return nodes.stream()
				.filter { node -> id == node.id }
				.findAny()
	}

	fun getEdge(start: Node, end: Node): Optional<LineSegment> {
		return if (start == end) {
			empty()
		} else edges.stream()
				.filter { edge -> edge.contains(start) }
				.filter { edge -> edge.contains(end) }
				.findAny()
	}
}
